Merge normal and important declarations in style rules

SimonSapin opened the pull request:
Have a single Vec instead of two. Fix #3426

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #3426.

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12943)

<!-- Reviewable:end -->

SimonSapin committed:
Add some fmt::Debug implementations

SimonSapin committed:
Add an Importance enum replacing booleans to indicate `!important`.

SimonSapin committed:
Remove unused Keyframe::parse method

SimonSapin committed:
Make parsing of keyframe declaration blocks spec-compliant.

Exclude animation properties.

SimonSapin committed:
Typo fixes

SimonSapin committed:
Unreverse declarations in memory.

SimonSapin committed:
Import SelectorMap back from the selectors crate.

Nobody else uses it, and I want to make breaking changes to it.

SimonSapin committed:
Sort rules by key in selector matching

SimonSapin committed:
Make SelectorMap and friends concrete

SimonSapin committed:
Merge normal and important declarations in style rules.

Have a single Vec instead of two. Fix #3426

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/selector_matching.rs, components/style/properties/properties.mako.rs, components/style/matching.rs, components/style/Cargo.toml, components/style/lib.rs, components/style/properties/data.py, components/style/properties/shorthand/box.mako.rs, components/style/dom.rs, components/style/keyframes.rs, components/style/properties/longhand/box.mako.rs, components/style/animation.rs
 * @kichjang: components/script/dom/cssstyledeclaration.rs, components/script/dom/element.rs, components/script_layout_interface/wrapper_traits.rs, components/script/layout_wrapper.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 2f500e9 with merge 95b3287...

bors-servo commented:
:snowman: The build was interrupted to prioritize another pull request.

bors-servo commented:
:hourglass: Trying commit 2f500e9 with merge 0095764...

SimonSapin committed:
Only cascade at a priority level rules that have declarations of that priority.

bors-servo commented:
:boom: Test timed out

emilio commented on components/style/keyframes.rs:
```diff
@@ -70,28 +72,7 @@ impl KeyframeSelector {
-            declarations: declarations.normal,
-        })
-    }
+    pub declarations: Arc<Vec<(PropertyDeclaration, Importance)>>,
```
Important declarations in keyframes are ignored, so the previous implementation was correct. Please make this just a `Vec<PropertyDeclaration>`, and point to https://drafts.csswg.org/css-animations/#keyframes:

> None of the properties interact with the cascade (so using !important on them is invalid and will cause the
> property to be ignored).

SimonSapin replied:
`style::animation::compute_style_for_animation_step` crates a `style::selector_matching::DeclarationBlock` struct from this. That struct contains `Arc<Vec<(PropertyDeclaration, Importance)>>`. So having dummy `Importance::Normal` values in `Keyframe` enables doing this by cloning an `Arc` (incrementing a refcount) rather than creating a new `Vec`.

I’ve added a doc-comment saying this.

emilio commented on components/style/keyframes.rs:
```diff
@@ -101,7 +82,7 @@ impl Keyframe {
 #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
 pub enum KeyframesStepValue {
-    Declarations(Arc<Vec<PropertyDeclaration>>),
+    Declarations(Arc<Vec<(PropertyDeclaration, Importance)>>),
```
ditto

SimonSapin replied:
Also added a comment here.

emilio commented on components/style/keyframes.rs:
```diff
@@ -259,12 +240,48 @@ impl<'a> QualifiedRuleParser for KeyframeListParser<'a> {
+        let mut iter = DeclarationListParser::new(input, parser);
+        while let Some(declaration) = iter.next() {
+            match declaration {
+                Ok(d) => declarations.extend(d.into_iter().map(|d| (d, Importance::Normal))),
```
Ok, nvm, this happens for reviewing top to bottom :-)

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -561,21 +560,20 @@ pub fn parse_property_declaration_list(context: &ParserContext, input: &mut Pars
     let mut deduplicated = vec![];
     let mut seen = PropertyBitField::new();
     let mut seen_custom = Vec::new();
     for declaration in declarations.into_iter().rev() {
```
Hmm... should this deduplicate in terms of both the property and the importance? It looks to me like it should, at least to match the previous behavior. If not, care to explain why? Because I'd be very confused.

SimonSapin replied:
Oh, that’s a good point. I’ll change this to eliminate a normal declaration if there’s an important one of the same name (regardless of order).

emilio commented on components/style/keyframes.rs:
```diff
@@ -259,12 +240,48 @@ impl<'a> QualifiedRuleParser for KeyframeListParser<'a> {
+        match PropertyDeclaration::parse(name, self.context, input, &mut results, true) {
+            PropertyDeclarationParseResult::ValidOrIgnoredDeclaration => {}
+            _ => return Err(())
+        }
```
Also, this will not ignore an important declaration, right? But convert it silently into a non-important one. Seems we should try to parse `!important` here, and return an error if we parse it.

emilio replied:
Nevermind, I see how this works now. I missed the `parse_until_before` on the normal DeclarationParser.

SimonSapin replied:
More importantly, `parse_important` is called in `PropertyDeclarationParser` but not here. And `parse_entirely` is called somewhere, to return `Err(())` if there’s stuff left unparsed at the end of the declaration value.

emilio commented on components/style/selector_matching.rs:
```diff
@@ -574,3 +583,318 @@ impl PerPseudoElementSelectorMap {
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct SelectorMap {
+    // TODO: Tune the initial capacity of the HashMap
+    id_hash: HashMap<Atom, Vec<Rule>>,
```
The rust-selectors hashmap used fnv if I'm not wrong. This would switch us to use sip-hash, and I'd expec this to be a severe perf impact. Can you switch it back to fnv?

SimonSapin replied:
Good catch. I’ve also changed `Stylist::animations`.

emilio commented on components/style/selector_matching.rs:
```diff
@@ -574,3 +583,318 @@ impl PerPseudoElementSelectorMap {
+/// element name, etc. will contain the Rules that actually match that
+/// element.
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct SelectorMap {
```
Also, it seems to me we could move this to its own file (maybe a `style/selector_matching` submodule?). Though I'd say that's fine as a followup.

SimonSapin replied:
Maybe. Though `Stylist` / `PerPseudoElementSelectorMap` / `PerOriginSelectorMap` / `PerOriginSelectorMap` / `SelectorMap` form a kind of hierarchy and we could cut at any point.

emilio commented on tests/unit/script/size_of.rs:
```diff
@@ -40,10 +40,10 @@ macro_rules! sizeof_checker (
-sizeof_checker!(size_htmlelement, HTMLElement, 352);
-sizeof_checker!(size_div, HTMLDivElement, 352);
-sizeof_checker!(size_span, HTMLSpanElement, 352);
+sizeof_checker!(size_element, Element, 328);
```
Sweet :-)

SimonSapin replied:
Looks like I hadn’t pushed my last commit which reverts this. (It adds a couple `bool` flags but alignment rounds to 8 bytes) Sorry :)

emilio commented:
Arguably should have reviewed commit by commit to avoid leaving so much noise, but it's probably late for that.

The approach looks good to me. Probably @heycam may want to take a look at this too. I'm still confused by the `deduplicate_property_declarations` thing, but the rest makes sense to me.

@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 2f500e9 with merge 7ed02d2...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2759)

highfive commented:
▶ FAIL [expected PASS] /css-style-attr-1_dev/html4/style-attr-cascade-002.htm
      └   → /css-style-attr-1_dev/html4/style-attr-cascade-002.htm bf25a67d9f07765e90c077aec3051e5effa9ed08
    /css-style-attr-1_dev/html4/reference/ref-green.htm 5570f2b94aae1cac828af0c91bc632190eb0f0c1
    Testing bf25a67d9f07765e90c077aec3051e5effa9ed08 == 5570f2b94aae1cac828af0c91bc632190eb0f0c1
    
      ▶ FAIL [expected PASS] /css-variables-1_dev/html/variable-reference-13.htm
      └   → /css-variables-1_dev/html/variable-reference-13.htm bf1aa2f1153c018c97c7fdd1ac77c71ff66e899d
    /css-variables-1_dev/html/support/color-green-ref.htm 90de59877a1f34c922b370c162008e33df4c6f30
    Testing bf1aa2f1153c018c97c7fdd1ac77c71ff66e899d == 90de59877a1f34c922b370c162008e33df4c6f30
    
      ▶ FAIL [expected PASS] /css21_dev/html4/style-attr-cascade-002.htm
      └   → /css21_dev/html4/style-attr-cascade-002.htm bf25a67d9f07765e90c077aec3051e5effa9ed08
    /css21_dev/html4/reference/ref-green.htm 5570f2b94aae1cac828af0c91bc632190eb0f0c1
    Testing bf25a67d9f07765e90c077aec3051e5effa9ed08 == 5570f2b94aae1cac828af0c91bc632190eb0f0c1

SimonSapin committed:
Use FNV hash in SelectorMap

SimonSapin commented:
I’ve pushed `!fixup` commits, separate for review but to be auto-squashed before merging.

These last three test failures were not intermittent. I could reliably reproduce them locally, and fixing the deduplication vs importance bug that you caught fixed them.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 7b0a881 with merge e3c563c...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2604)

SimonSapin commented:
Rebased and fixed up for types renamed in https://github.com/servo/servo/pull/12947

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 5c2b1de with merge f285194...

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12861) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin committed:
Import SelectorMap unit tests from the selectors crate.

SimonSapin committed:
Revert "Simplify CSSStyleDeclaration::GetPropertyValue"

This reverts commit 1637b0ba8a66839e3329a3cf552ab2d4fb3c67c6.

* As far as I know, `fn` pointers don’t necessarily inline well
* Upcoming commits are gonna change this mapping to be less trivial,
  so this would at least need a new `fn` declaration, making this
  less of a simplification.

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit da33bc3 with merge 8e538ce...

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -596,40 +596,74 @@ pub fn parse_property_declaration_list(context: &ParserContext, input: &mut Pars
+                                continue
+                            }
+                            if seen_normal.get_${property.ident}() {
+                                remove_one(&mut deduplicated, |d| {
```
Hmmm... I was wondering how common was to arrive here, and what's the normal length of the `deduplicated` vector, but it seems in average this is rather uncommon, and also not a huge deal.

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -596,40 +596,74 @@ pub fn parse_property_declaration_list(context: &ParserContext, input: &mut Pars
+fn remove_one<T, F: FnMut(&T) -> bool>(v: &mut Vec<T>, mut remove_this: F) {
+    let previous_len = v.len();
+    v.retain(|x| !remove_this(x));
+    assert_eq!(v.len(), previous_len - 1);
```
`debug_assert!`, maybe?

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -588,18 +588,24 @@ pub fn parse_property_declaration_list(context: &ParserContext, input: &mut Pars
 /// The input and output are in source order
 fn deduplicate_property_declarations(declarations: Vec<(PropertyDeclaration, Importance)>)
-                                     -> Vec<(PropertyDeclaration, Importance)> {
+                                     -> (Vec<(PropertyDeclaration, Importance)>, bool) {
```
hmm... Maybe this would be clearer just comparing the input with the output length? Also, note that we don't need to recalc everything, we can be a bit smarter. If there was any important declaration, `any_important` would still always be true. And if there wasn't any important one, and `any_normal` was true, `any_normal` still needs to be true.

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -287,6 +287,28 @@ impl Importance {
+    pub fn recalc_any(&mut self) {
+        let mut any_normal = false;
+        let mut any_important = false;
+        for &(_, importance) in &*self.declarations {
```
also, we can short-circuit here if any_important and any_normal are both true, with that and the previous comment we should make faster nearly all cases.

SimonSapin replied:
As discussed on IRC, I removed this method entirely by maintaining a pair of `u32` counts (that can be decremented) instead of just `bool` presence flags.

emilio commented on components/style/selector_matching.rs:
```diff
@@ -609,16 +609,16 @@ impl PerPseudoElementSelectorMap {
-    id_hash: FnvHashMap<Atom, Vec<Rule>>,
-    class_hash: FnvHashMap<Atom, Vec<Rule>>,
-    local_name_hash: FnvHashMap<Atom, Vec<Rule>>,
+    pub id_hash: FnvHashMap<Atom, Vec<Rule>>,
```
Ugh, I'd prefer us not doing all this public, but I can't find a better way really... Maybe accessors that are prefixed with `test_only` or something?

SimonSapin replied:
Yeah, this is unfortunate but the same has happened in many other cases. As long as tests are not in the same crate, things used by tests have to be public. On the other hand having in-crate unit tests doubles the compile time for that crate. (You have to build everything again with `--test`.)

SimonSapin replied:
Accessors would work, but they’re not worth it IMO.

emilio replied:
Fair enough.

SimonSapin committed:
Keep track of the number of important and normal declarations in a block

… instead of the presence (`bool` flags) of each.

This allows removing `recalc_any` which iterated over the `Vec`.

emilio commented:
r=me when it's ready, @SimonSapin. As discussed, we can only keep track of the important declarations, and use the total length for computing the normal one. But I'm fine leaving that as a followup E-Easy issue.

emilio commented:
@bors-servo: R+

bors-servo commented:
:pushpin: Commit 023a3b7 has been approved by `emilio`

<!-- @bors-servo r=emilio 023a3b79912344ac6e5900171514890c73d6534c -->

bors-servo commented:
:hourglass: Testing commit 023a3b7 with merge 87fd6f1...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2779)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/setpropertypriority.html
      └   → /_mozilla/css/setpropertypriority.html 1c40214a1a913cbd096a44b2d280fb3193fa09d0
    /_mozilla/css/setpropertypriority_ref.html f565a85a358e1e10906e0a87eb234acff04e9bfb
    Testing 1c40214a1a913cbd096a44b2d280fb3193fa09d0 == f565a85a358e1e10906e0a87eb234acff04e9bfb

emilio commented on components/script/dom/element.rs:
```diff
@@ -780,81 +787,76 @@ impl Element {
+                let declarations = Arc::make_mut(&mut declarations.declarations);
+                for &mut (ref declaration, ref mut importance) in declarations {
+                    if properties.iter().any(|p| declaration.name() == **p) {
+                        *importance = new_importance
```
forgot to update the count in this one, that's why that test is failing, sorry for not catching it before.

SimonSapin commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit f9150af has been approved by `emilio`

<!-- @bors-servo r=emilio f9150af9367c9aff7184010b0656fd2881eae7cb -->

bors-servo commented:
:hourglass: Testing commit f9150af with merge f8b2be1...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2557), [arm64](http://build.servo.org/builders/arm64/builds/2570), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2621), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2782), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2642), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2629), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2724), [windows-dev](http://build.servo.org/builders/windows-dev/builds/517)
